carl/envs/carl_env_new.py

```diff
@@ -0,0 +1,253 @@
+        )
+
+    def get_observation_space(
+        self, obs_context_feature_names: list[str] | None = None
+    ) -> gymnasium.spaces.Dict:
+        context_space = self.get_context_space()
+        obs_space_context = context_space.to_gymnasium_space(
+            context_feature_names=obs_context_feature_names,
+            as_dict=self.obs_context_as_dict,
+        )
+        obs_space = spaces.Dict(
+            {
+                "state": self.state_observation_space,
+                "context": obs_space_context,
+            }
+        )
+        return obs_space
+
+    @staticmethod
+    def get_context_features() -> list[ContextFeature]:
+        # Child env specific
+        raise NotImplementedError
+
+    @classmethod
+    def get_context_space(cls) -> ContextSpace:
+        return ContextSpace(cls.get_context_features())
+
+    @classmethod
+    def get_default_context(cls) -> dict[str, Any]:
+        default_context = cls.get_context_space().get_default_context()
+        return default_context
+
+    def _progress_instance(self) -> None:
+        """
+        Progress instance.
+
+        In this case instance is a specific context.
+        1. Select instance with the instance_mode. If the instance_mode is random, randomly select
+        the next instance from the set of contexts. If instance_mode is rr or roundrobin, select
+        the next instance.
+        2. If Gaussian noise should be added to whitelisted context features, do so.
+
+        Returns
+        -------
+        None
+
+        """
+        context = self.context_selector.select()  # type: ignore [call-arg]
+        self.context = context
+
+    def reset(
+        self, *, seed: int | None = None, options: dict[str, Any] | None = None
+    ) -> tuple[Any, dict[str, Any]]:
+        self._progress_instance()
+        self._update_context()
+        state = super().reset(seed=seed, options=options)
+        state = self._add_context_to_state(state)
+        return state
+
+    def _add_context_to_state(self, state):
+        context = self.context
+        if not self.obs_context_as_dict:
+            context = [self.context[k] for k in self.state_context_features]
+        else:
+            context = {
+                k: v
+                for k, v in self.context.items()
+                if k in self.state_context_features
+            }
+        state_context_dict = {
+            "state": state,
+            "context": context,
+        }
+        return state_context_dict
+
+    @abc.abstractmethod
+    def _update_context(self) -> None:
+        """
+        Update the context feature values of the environment.
+
+        Returns
+        -------
+        None
+
+        """
+        ...
+
+    def step(
+        self, action: Any
+    ) -> tuple[Any, SupportsFloat, bool, bool, dict[str, Any]]:
+        state = super().step(action)
+        state = self._add_context_to_state(state)
+        return state
+
+
+class CARLClassicControlEnv(CARLEnv):
```

Comment:
I might overlook something, but why do we want/need a separate subclass here? If we want the gym.make, I'd suggest having a CARLGymEnv instead of a subenv per domain

Reply:
Yes you are right! Makes it even prettier
- [ ] Create `CARLGymEnv`